File: sdks/typescript/src/connect-proto.ts
import { HttpClient, HTTPError } from "./http-client.js";

const CONNECT_PROTOCOL_VERSION = "Connect-Protocol-Version";

export async function unaryPost<T>(
  hc: HttpClient,
  url: string,
  headers: Record<string, string>,
  reqMsg: unknown,
): Promise<T> {
  const h = { ...headers, [CONNECT_PROTOCOL_VERSION]: "1" };
  try {
    const resp = await hc.doRaw(
      "POST",
      url,
      h,
      JSON.stringify(reqMsg),
      "application/json",
    );
    if (resp.body.length === 0) return undefined as T;
    const text = new TextDecoder().decode(resp.body);
    const parsed = JSON.parse(text);
    if (parsed.error) {
      throw new Error(
        `connect error: ${parsed.error.message || JSON.stringify(parsed.error)}`,
      );
    }
    if (parsed.result !== undefined) return parsed.result as T;
    return parsed as T;
  } catch (e) {
    if (e instanceof HTTPError) {
      throw decodeConnectErr(e);
    }
    throw e;
  }
}

export async function streamPost(
  hc: HttpClient,
  url: string,
  headers: Record<string, string>,
  reqMsg: unknown,
  each: (msg: unknown) => void,
): Promise<void> {
  const h: Record<string, string> = {
    ...headers,
    [CONNECT_PROTOCOL_VERSION]: "1",
    "Content-Type": "application/json",
    Accept: "application/connect+json",
  };
  const resp = await hc.fetchRaw(url, {
    method: "POST",
    headers: h,
    body: JSON.stringify(reqMsg),
  });
  if (!resp.body) throw new Error("no response body for stream");
  const reader = resp.body.getReader();
  let buffer = new Uint8Array(0);

  while (true) {
    const { done, value } = await reader.read();
    if (value) {
      const newBuf = new Uint8Array(buffer.length + value.length);
      newBuf.set(buffer);
      newBuf.set(value, buffer.length);
      buffer = newBuf;
    }
    // Process complete frames
    while (buffer.length >= 5) {
      const frameLen = new DataView(
        buffer.buffer,
        buffer.byteOffset + 1,
        4,
      ).getUint32(0);
      if (frameLen <= 0 || frameLen > 64 * 1024 * 1024) {
        throw new Error(`connect stream: invalid frame size ${frameLen}`);
      }
      if (buffer.length < 5 + frameLen) break;
      const payload = buffer.slice(5, 5 + frameLen);
      buffer = buffer.slice(5 + frameLen);
      const text = new TextDecoder().decode(payload);
      const parsed = JSON.parse(text);
      if (parsed.error) {
        throw new Error(
          `connect stream error: ${parsed.error.message || JSON.stringify(parsed.error)}`,
        );
      }
      each(parsed);
    }
    if (done) break;
  }
}

function decodeConnectErr(he: HTTPError): Error {
  try {
    const text = new TextDecoder().decode(he.body);
    const env = JSON.parse(text);
    if (env.message) return new Error(`http ${he.status}: ${env.message}`);
  } catch {
    /* ignore */
  }
  return new Error(`http ${he.status}: ${new TextDecoder().decode(he.body)}`);
}

File: sdks/python/src/sandboxer/_connect_proto.py
from __future__ import annotations

import json
import struct
from typing import Callable, Any

from ._http_client import HttpClient, AsyncHttpClient, HTTPError

CONNECT_PROTOCOL_VERSION = "Connect-Protocol-Version"
CONNECT_TIMEOUT_MS = "Connect-Timeout-Ms"


def unary_post(hc: HttpClient, url: str, headers: dict[str, str], req_msg: Any) -> Any:
    h = {**headers, CONNECT_PROTOCOL_VERSION: "1"}
    try:
        status, raw = hc.do_raw("POST", url, h, content=json.dumps(req_msg).encode(), content_type="application/json")
    except HTTPError as e:
        raise _decode_connect_err(e) from e
    if not raw:
        return None
    parsed = json.loads(raw)
    if "error" in parsed and parsed["error"]:
        msg = parsed["error"].get("message", json.dumps(parsed["error"]))
        raise Exception(f"connect error: {msg}")
    if "result" in parsed:
        return parsed["result"]
    return parsed


def stream_post(hc: HttpClient, url: str, headers: dict[str, str], req_msg: Any, each: Callable[[Any], None]) -> None:
    h = {
        **headers,
        CONNECT_PROTOCOL_VERSION: "1",
        "Content-Type": "application/json",
        "Accept": "application/connect+json",
    }
    resp = hc.request("POST", url, h, content=json.dumps(req_msg).encode())
    if resp.status_code >= 400:
        raise _decode_connect_err(HTTPError(resp.status_code, resp.content))
    _read_connect_stream(resp.content, each)


def _read_connect_stream(data: bytes, each: Callable[[Any], None]) -> None:
    offset = 0
    while offset + 5 <= len(data):
        frame_len = struct.unpack(">I", data[offset + 1 : offset + 5])[0]
        if frame_len <= 0 or frame_len > 64 * 1024 * 1024:
            raise Exception(f"connect stream: invalid frame size {frame_len}")
        if offset + 5 + frame_len > len(data):
            break
        payload = data[offset + 5 : offset + 5 + frame_len]
        offset += 5 + frame_len
        parsed = json.loads(payload)
        if "error" in parsed and parsed["error"]:
            msg = parsed["error"].get("message", json.dumps(parsed["error"]))
            raise Exception(f"connect stream error: {msg}")
        each(parsed)


def _decode_connect_err(he: HTTPError) -> Exception:
    try:
        env = json.loads(he.body)
        if "message" in env:
            return Exception(f"http {he.status}: {env['message']}")
    except (json.JSONDecodeError, KeyError):
        pass
    return Exception(f"http {he.status}: {he.body.decode(errors='replace')}")
